### PR TITLE
Implement production entity change request

### PR DESCRIPTION
**Pull request by MKodde**
Leveraging the work from #516, we are now able to perform entity change requests on manage for production entities. Those entities now have an edit button. Clicking that takes the user to a regular edit form. The form results are dropped on the change request Manage endpoint. Resulting in a entity change request on Manage, and a custom thank you page for the end user.

The following features have not yet been implemented like:
- Create Jira ticket
- Send confirmation mail message
Although some provisional work to implement those features have been included in this PR (copy paste bonus)

See: https://www.pivotaltracker.com/story/show/182411730

### CODE REVIEW

File: src/Surfnet/ServiceProviderDashboard/Domain/ValueObject/Contact.php [hunk 1]
```diff
@@ -168,4 +168,14 @@ public function isContactSet(): bool
     {
         return !(empty($this->firstName) && empty($this->lastName) && empty($this->email) && empty($this->phone));
     }
+
+    public function isSet()
+    {
+        $isNull = $this->email === null &&
+        $this->firstName === null &&
```
Comment (Shaky212) [inline src/Surfnet/ServiceProviderDashboard/Domain/ValueObject/Contact.php:175]: Indent 3 lines to make it more readable.

Comment (MKodde) [reply src/Surfnet/ServiceProviderDashboard/Domain/ValueObject/Contact.php:175]: This looks to be a regression from the previous PR. the isSet method was removed in favor of the already present 'isContactSet'. Will remove this entirily

File: src/Surfnet/ServiceProviderDashboard/Application/CommandHandler/Entity/EntityChangeRequestCommandHandler.php [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+        $this->issueType = $issueType;
+        $this->summaryTranslationKey = 'entity.publish.change_request.ticket.summary';
+        $this->descriptionTranslationKey = 'entity.publish.change_request.ticket.description';
+    }
+
+    /**
+     * Creates an entity change request in Manage
+     * @SuppressWarnings(PHPMD.ElseExpression)
+     */
+    public function handle(PublishProductionCommandInterface $command)
```
Comment (johanib) [inline src/Surfnet/ServiceProviderDashboard/Application/CommandHandler/Entity/EntityChangeRequestCommandHandler.php:120]: A new class with 2 ignores. Are these justified?
The handle function is quite big. Does it adhere to the SingleResponsibilityPrinciple?

File: src/Surfnet/ServiceProviderDashboard/Application/ViewObject/EntityActions.php [hunk 3]
```diff
@@ -86,8 +86,7 @@ public function getEnvironment()
     public function allowEditAction(): bool
     {
         $notEditable =
-            $this->isPublishedToProduction()
-            || $this->readOnly
+            $this->readOnly
```
Comment (johanib) [inline src/Surfnet/ServiceProviderDashboard/Application/ViewObject/EntityActions.php:89]: Random question: Is this change tested?

File: src/Surfnet/ServiceProviderDashboard/Infrastructure/DashboardBundle/Controller/EntityControllerTrait.php [hunk 4]
```diff
@@ -140,18 +141,23 @@ private function handleImport(Request $request, SaveSamlEntityCommand $command)
...
-        switch ($entity->getEnvironment()) {
-            case Constants::ENVIRONMENT_TEST:
+        switch (true) {
+            case $entity->getEnvironment() === Constants::ENVIRONMENT_TEST:
                 $publishEntityCommand = new PublishEntityTestCommand($entity);
                 $destination = 'entity_published_test';
                 break;
-
-            case Constants::ENVIRONMENT_PRODUCTION:
+            case $isEntityChangeRequest:
```
Comment (johanib) [inline src/Surfnet/ServiceProviderDashboard/Infrastructure/DashboardBundle/Controller/EntityControllerTrait.php:155]: err, this is weird? :thinking: 
Should this not be an if before the switch?

File: src/Surfnet/ServiceProviderDashboard/Infrastructure/DashboardBundle/Resources/config/services.yml [hunk 5]
```diff
@@ -88,6 +88,13 @@ services:
         tags:
             - { name: tactician.handler, command: Surfnet\ServiceProviderDashboard\Application\Command\Entity\PublishEntityTestCommand }
 
+    Surfnet\ServiceProviderDashboard\Application\CommandHandler\Entity\EntityChangeRequestCommandHandler:
+        arguments:
+            $repository: '@Surfnet\ServiceProviderDashboard\Infrastructure\Manage\Client\EntityChangeRequestClient'
+            $issueType: 'TODO create configurable parameter'
```
Comment (johanib) [inline src/Surfnet/ServiceProviderDashboard/Infrastructure/DashboardBundle/Resources/config/services.yml:94]: Is this something you should do, or just the value of the $issueType??

**Review by johanib — COMMENTED** [review]
Looks fine in general, some remarks.

File: src/Surfnet/ServiceProviderDashboard/Application/ViewObject/EntityActions.php (hunk 3 above)
Comment (MKodde) [reply src/Surfnet/ServiceProviderDashboard/Application/ViewObject/EntityActions.php:89]: Albeit a small change of an existing test, yes it was covered in test here: `tests/integration/Application/ViewObject/EntityActionsTest.php:144`

File: src/Surfnet/ServiceProviderDashboard/Application/CommandHandler/Entity/EntityChangeRequestCommandHandler.php (hunk 2 above)
Comment (MKodde) [reply src/Surfnet/ServiceProviderDashboard/Application/CommandHandler/Entity/EntityChangeRequestCommandHandler.php:120]: This one's a pickle. This command handler is tasked with many things indeed. Maybe this is a good time to extract some of its (shared) logic into services of their own.

Thanks for the suggestion! I will add this refactor to the #523 that branched of this work. That prevents me to rebase/rework this too much.

File: src/Surfnet/ServiceProviderDashboard/Application/ViewObject/EntityActions.php (hunk 3 above)
Comment (johanib) [reply src/Surfnet/ServiceProviderDashboard/Application/ViewObject/EntityActions.php:89]: :100:

File: src/Surfnet/ServiceProviderDashboard/Infrastructure/DashboardBundle/Controller/EntityControllerTrait.php (hunk 4 above)
Comment (MKodde) [reply src/Surfnet/ServiceProviderDashboard/Infrastructure/DashboardBundle/Controller/EntityControllerTrait.php:155]: _Some background information for what its worth:_

This switch determines what entity edit action should take place. 

Are we editing a test entity, a production entity or maybe edit a published production entity. Previously all of these decisions could be made based on the publication state of the entity. For this new 'change request', which is basically an edit on a published entity, there was need for an additional check. 

To me this is not weird. Before I switched on the entity status. Now that we have two 'hits' on a published entity, I had to add a condition to the switch condition. And my go to trick is to move the conditions to the case statements. And switch on `true`.

Or is this not what you find weird?

If you like, I can explain IRL

File: src/Surfnet/ServiceProviderDashboard/Infrastructure/DashboardBundle/Resources/config/services.yml (hunk 5 above)
Comment (MKodde) [reply src/Surfnet/ServiceProviderDashboard/Infrastructure/DashboardBundle/Resources/config/services.yml:94]: This was a temporary reminder to change this to the actual issueType. In #523 this is fixed permanently.

_Background information_
I yanked code from the `PublishEntityProductionCommandHandler`. As the handling of this command is performed in a very similar fashion. For the time being, I left the Jira implementation of this command handler unchanged. As that was to be addressed in a successive story (see #523).

Given your previous comment. I'm planning on extracting some of the shared logic from these two command handlers, and put them in a service.

**Comment by MKodde** [comment]
This story was implicitly approved in #522

Merging...

**Pull request merged**